icu4c/source/common/uloc.cpp

```diff
@@ -1465,6 +1465,25 @@ uloc_openKeywords(const char* localeID,
 static const char i_default[] = {'i', '-', 'd', 'e', 'f', 'a', 'u', 'l', 't'};
 #define I_DEFAULT_LENGTH UPRV_LENGTHOF(i_default)
 
+static int32_t _preprocessLocaleID(const char* inLocaleID, char* outBuffer, int32_t bufferCapacity, UErrorCode* err) {
+    // this function is to allow permissive parsing in locale IDs in general cases, where the locale ID can be in BCP
+    // 47 format, but with underbars as field delimiters
+    int32_t neededLength = static_cast<int32_t>(uprv_strlen(inLocaleID));
+    if (neededLength > bufferCapacity) {
+        *err = U_BUFFER_OVERFLOW_ERROR;
+        return neededLength;
+    }
+    
+    uprv_strcpy(outBuffer, inLocaleID);
+    if (uprv_strchr(inLocaleID, '_') != nullptr && ((inLocaleID[0] != 'x' && inLocaleID[0] != 'X') || (inLocaleID[0] != '\0' && inLocaleID[1] != '-'))) {
+        // convert all underbars to hyphens, unless the string starts with "x-" and contains an underbar
```

Comment:
What if it starts with "x_" ?

Reply:
This was driven by what the unit tests were telling me, and we didn't have any failing unit tests with an underbar in the second position.  Only hyphens.  So I concluded that there was a syntax rule that required that separator to be a hyphen.

However, in trying to get my revised version of this code working correctly, I encountered test failures where the first letter could be something other than "x" or "X", and where the separator could, indeed be an underbar.  I decided to loosen the condition here so that it just checks to see if the second character in the locale ID is `_` or `-`.  This seems to work without apparent side effects.  I'll make the same change in the Java version.

Reply:
This looks better now, but I still don't quite understand the conditions.
- "x-some-thing" is a valid BCP 47 language tag, and I think that writing it with underscores should work just as well as "de_AT-u-kn".
- "u-some-thing" is not valid BCP 47, nor really meaningful in legacy ICU syntax; we probably don't care how this is handled.

So... are there test cases (that look meaningful) which fail if you *only* check for `uprv_strchr(localeID, '_')` ?

Reply:
Oh, I think you mention this in the PR description. Without this, "x-klingon" becomes "@x=klingon".
But... @yumaoka isn't that what we would expect?

Reply:
All I can say here is that we had a bunch of unit tests that were specifically expecting us not to mess with "x-klingon" and friends.  I can't cite the specific ones off the top of my head, but if you want, I can back out that change and let you know specifically what failed.

Reply:
I am letting the "x-klingon" question slide. Not important enough to block this good work.